sled-agent/src/instance_manager.rs

```diff
@@ -775,6 +800,53 @@ impl InstanceManagerRunner {
         tx.send(Ok(state)).map_err(|_| Error::FailedSendClientClosed)?;
         Ok(())
     }
+
+    async fn only_use_disks(
+        &mut self,
+        tx: oneshot::Sender<Result<(), Error>>,
+        disks: AllDisks,
+    ) -> Result<(), Error> {
+        // Consider the generation number on the incoming request to avoid
+        // applying old requests.
+        let requested_generation = *disks.generation();
+        if let Some(last_gen) = self.storage_generation {
+            if last_gen >= requested_generation {
+                // This request looks old, ignore it.
+                info!(self.log, "only_use_disks: Ignoring request";
+                    "last_gen" => ?last_gen, "requested_gen" => ?requested_generation);
+
+                return Ok(());
+            }
+        }
+        self.storage_generation = Some(requested_generation);
+        info!(self.log, "only_use_disks: Processing new request";
+            "gen" => ?requested_generation);
+
+        let u2_set: HashSet<_> = disks.all_u2_zpools().into_iter().collect();
+
+        let mut to_remove = vec![];
+        for (id, (_, instance)) in self.instances.iter() {
+            // If we can read the filesystem pool, consider it. Otherwise, move
+            // on, to prevent blocking the cleanup of other instances.
+            let Ok(Some(filesystem_pool)) =
+                instance.get_filesystem_zpool().await
+            else {
+                info!(self.log, "only_use_disks: Cannot read filesystem pool"; "id" => ?id);
+                continue;
+            };
+            if !u2_set.contains(&filesystem_pool) {
+                to_remove.push(*id);
+            }
+        }
+
+        for id in to_remove {
+            info!(self.log, "only_use_disks: Removing instance"; "id" => ?id);
+            self.instances.remove(&id);
```

Comment:
What actually happens when we remove an instance from this map? What happens to the instance as a result? 

Per the comment in `sled_agent.rs`, it seems like we _want_ to mark as failed any instances which are using expunged disks, but I don't actually see where that happens.

Reply:
Great question - we don't actually explicitly mark the instance failed here, we just stop it from running.

This is basically the same pathway a VMM could use to terminate itself.
This is the object we end up removing via the call to `self.instances.remove`:

https://github.com/oxidecomputer/omicron/blob/30b6713305faa57e80b25089a66c0b99f0553ca7/sled-agent/src/instance.rs#L904-L910

And by dropping it, we remove the only sender-side of an `mpsc` of `InstanceRequest`s.

This should trigger the self-terminating case of the `InstanceRunner`:

https://github.com/oxidecomputer/omicron/blob/30b6713305faa57e80b25089a66c0b99f0553ca7/sled-agent/src/instance.rs#L450-L454

Reply:
This is kinda why I mentioned I'd be relying on the background task -- as implemented, I'd need to wait for a background task to notice that the instance is not running on the sled.

I *can* introduce an upcall to nexus, but I figured that could throw more of a wrench into the instance lifecycle revamp. Lemme know what you prefer!

Reply:
Coming back to this after a couple days, I guess it's possible that we return from the `omicron_physical_disks_ensure` function before the instance runner has finished terminating. I'll await the join handle here too.

Reply:
Done in https://github.com/oxidecomputer/omicron/pull/5965/commits/e360daefbf3409b6df795f0e01ca45f543ff4f43

Reply:
> we don't actually explicitly mark the instance failed here, we just stop it from running.

That's good! Eventually, sled-agent will never mark instances failed. I think we _do_ want to mark the _vmm_ state as failed, though.

Reply:
Ack, thanks for bearing with me. I'm giving this a shot in: https://github.com/oxidecomputer/omicron/pull/5965/commits/f242e0aff03129af7d93e87b9344420fb7f2f813

Right now, I'm just passing an extra boolean to some of the termination functions to decide what VMM state to propagate. Later, I think it might be useful to expand this to a stringified message too, so we can gain some better diagnostic info on **why** an instance failed or stopped.